tests/Elastic.Transport.Tests/ResponseBuilderDisposeTests.cs

```diff
@@ -3,146 +3,70 @@
 // See the LICENSE file in the project root for more information
 
 using System;
-using System.Collections.Generic;
 using System.IO;
-using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Elastic.Transport.Tests.Plumbing;
 using FluentAssertions;
 using Xunit;
 
-namespace Elastic.Transport.Tests
+namespace Elastic.Transport.Tests;
+
+public class ResponseBuilderDisposeTests
 {
-	public class ResponseBuilderDisposeTests
-	{
-		private readonly ITransportConfiguration _settings = InMemoryConnectionFactory.Create().DisableDirectStreaming(false);
-		private readonly ITransportConfiguration _settingsDisableDirectStream = InMemoryConnectionFactory.Create().DisableDirectStreaming();
+	private readonly ITransportConfiguration _settings = InMemoryConnectionFactory.Create().DisableDirectStreaming(false);
 
-		[Fact] public async Task ResponseWithHttpStatusCode() => await AssertRegularResponse(false, 1);
+	[Fact]
+	public async Task ResponseWithPotentialBody_StreamIsNotDisposed() => await AssertResponse(expectedDisposed: false);
 
-		[Fact] public async Task ResponseBuilderWithNoHttpStatusCode() => await AssertRegularResponse(false);
+	[Fact]
+	public async Task ResponseWith204StatusCode_StreamIsDisposed() => await AssertResponse(204);
 
-		[Fact] public async Task ResponseWithHttpStatusCodeDisableDirectStreaming() =>
-			await AssertRegularResponse(true, 1);
+	[Fact]
+	public async Task ResponseForHeadRequest_StreamIsDisposed() => await AssertResponse(httpMethod: HttpMethod.HEAD);
 
-		[Fact] public async Task ResponseBuilderWithNoHttpStatusCodeDisableDirectStreaming() =>
-			await AssertRegularResponse(true);
+	[Fact]
+	public async Task ResponseWithZeroContentLength_StreamIsDisposed() => await AssertResponse(contentLength: 0);
 
-		private async Task AssertRegularResponse(bool disableDirectStreaming, int? statusCode = null)
+	private async Task AssertResponse(int statusCode = 200, HttpMethod httpMethod = HttpMethod.GET, int contentLength = 10, bool expectedDisposed = true)
+	{
+		var settings = _settings;
+		var requestData = new RequestData(httpMethod, "/", null, settings, null, null, default)
 		{
-			var settings = disableDirectStreaming ? _settingsDisableDirectStream : _settings;
-			var memoryStreamFactory = new TrackMemoryStreamFactory();
```

Comment:
The purpose of this was also to ensure the memory stream factory returned memory streams get disposed properly. I think we should still test this. 

How should `DisableDirectStreaming: true` and `StreamResponse` behave?

We should have a matrix test over 

`DisableDirectStreaming` | `Type of Response`

That asserts both the stream of the response AND any memory streams return by the memory stream factory (the real implementation uses a shared pool so returns are very important)

Testing `StreamResponse` and `StringResponse` should suffice.

Reply:
Yeah, that's a good call out. I'll reintroduce something to cover these scenarios as well.